predicates/auth/headersha256.go

```diff
@@ -0,0 +1,76 @@
+package auth
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"net/http"
+
+	"github.com/zalando/skipper/predicates"
+	"github.com/zalando/skipper/routing"
+)
+
+type headerSha256Spec struct{}
+
+type headerSha256Predicate struct {
+	name   string
+	hashes [][]byte
+}
+
+// NewHeaderSHA256 creates a predicate specification, whose instances match SHA-256 hash of the header value.
+// The HeaderSHA256 predicate requires the header name and one or more hex-encoded SHA-256 hash values of the matching header.
+func NewHeaderSHA256() routing.PredicateSpec {
+	return &headerSha256Spec{}
+}
+
+func (*headerSha256Spec) Name() string {
+	return predicates.HeaderSHA256Name
+}
+
+// Create a predicate instance matching SHA256 hash of the header value
+func (*headerSha256Spec) Create(args []interface{}) (routing.Predicate, error) {
+	if len(args) < 2 {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	name, ok := args[0].(string)
+	if !ok {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	var hashes [][]byte
+	for _, arg := range args[1:] {
+		hexHash, ok := arg.(string)
+		if !ok {
+			return nil, predicates.ErrInvalidPredicateParameters
+		}
+		hash, err := hex.DecodeString(hexHash)
+		if err != nil {
+			return nil, err
+		}
+		if len(hash) != sha256.Size {
+			return nil, predicates.ErrInvalidPredicateParameters
+		}
+		hashes = append(hashes, hash)
+	}
+
+	return &headerSha256Predicate{name, hashes}, nil
+}
+
+func (p *headerSha256Predicate) Match(r *http.Request) bool {
+	value := r.Header.Get(p.name)
+	if value == "" {
+		return false
+	}
+
+	s := sha256.Sum256([]byte(value))
+	valueHash := s[:]
```

Comment:
why do we need to copy here?
I guess `[][]byte` iterate so you compare `[]byte` with `[Size]byte`, but maybe we can create `[][Size]`, `Size` is defined by `sha256.Size` so should be possible.
Do I miss something?

Reply:
its not a copy, it creates slice from array returned by sha256.Sum256

Reply:
> but maybe we can create [][Size]

but then we would need to convert arrays to slices anyway to compare :) https://stackoverflow.com/questions/18561219/comparing-arrays-in-go-language

Reply:
whatever you think is better. I think the preparation should always be the heavy lifting if possible and the hot path as few lines as possible, so I tend to the change, but if it's the same in benchmark results, then whatever you prefer seems fine